R/FilterState.R

```diff
@@ -575,15 +562,17 @@ FilterState <- R6::R6Class( # nolint
     # @description
     # Return variable name prefixed by `dataname` to be evaluated as extracted object,
     # for example `data$var`
+    # @param dataname `character(1)` name of a dataset
+    # @param extract_type `character(1)` type of extraction, one of `character(0)`, `"list"`, `"matrix"`
```

Comment:
- I kind of dislike mixing `character(0)` with `character(1)`. Usually there is a support of "none" value instead of `character(0)`.
- If you do the above you can then very easily validate values with `match.arg()`. Current implementation is rather: `"list"`, `"matrix"` and everything else. (don't know if this is checked elsewhere - I'm looking at this method only)

Reply:
I don't know exactly how to call each way of extraction:
- `character(0)` means that variables in a `subset` call are not prefixed (`data |> subset(x == "a")`)
- `list` means that variables are extracted by `$` in a subset call (`data |> subsetByColData(data$x == "a")`) 
- `matrix` means that variables are extracted by `[[` in a subset call (`data |> subset(data[, "x"] == "a")`)

Somebody has better terms for above cases?

```suggestion
    # @param extract_type `character(1)` type of extraction, one of `"nse"`, `"list"`, `"matrix"`
```

Reply:
Perhaps change the name of the argument to `var_prefix` or something like that and than `c("none", "list", "matrix")` will be proper options?

Reply:
> Perhaps change the name of the argument to `var_prefix` or something like that and than `c("none", "list", "matrix")` will be proper options?

Yes exactly.

I have one more thought but I'm not yet super convinced to this. I'm interested in hearing your thoughts on this.
Instead of a string which is then matched to sprintf pattern, we can allow providing sprintf pattern or the whole sprintf call.
Taking this idea further - we are actually talking about an expression and not message / warning / anything else that requires a character type of data. Hance, instead of a sptrintf that produced a char which is then parsed to expression -> move to substitute that returns expression directly without string based step. I mean - operate on the right data type.

Reply:
1. Note this is an internal class and the user never provides anything, unless you count senior classes as users.
2. We have gone through operating on strings and language objects throughout `teal.slice` back in Spring. Trust me, the code to build this as a string is much, much simpler than that required to build expression. This is currently (I believe) the only departure from building expressions from language objects in the package.

Reply:
> I have one more thought but I'm not yet super convinced to this. I'm interested in hearing your thoughts on this.
Instead of a string which is then matched to sprintf pattern, we can allow providing sprintf pattern or the whole sprintf call.

Theoretically you can just provide `get_call(varname = "dataname$varname")`, but this doesn't make sense, because `dataname` and `varname` is an information kept inside of `FilterState` class. When gathering calls from all `FilterState` object, parent (method which gathers) doesn't know anything about `varname`. 

Providing sprintf pattern is not enough neither, because:
1. `sprintf("%s$%s", dataname, varname)`
2. `sprintf('%s[, "%s"]', dataname, varname)`
3. `sprintf("%s", varname)

As you can see, dataname in first two is a first argument and varname second. In case of (3) dataname is not an argument for sprintf at all and varname is second.

Pattern is hard do handle because different arguments are taken in all cases.